File: C++/010. Regular Expression Matching.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    bool isMatch(std::string s, std::string p) {
        auto M = int(s.size());
        auto N = int(p.size());
        std::vector<std::vector<bool>> dp (N + 1, std::vector<bool>(M + 1, false));
        s = "*" + s;
        p = "*" + p;
        for (int i = 0; i <= N; i += 2) {
            if (p[i] == '*')
                dp[i][0] = true;
            else
                break;
        }
        for (int i = 1; i <= N; i++) {
            for (int j = 1; j <= M; j++) {
                if ((p[i] == s[j] || p[i] == '.') && dp[i - 1][j - 1])
                    dp[i][j] = true;
                if (p[i] == '*') {
                    if (dp[i - 2][j] == true)
                        dp[i][j] = true;
                    else {
                        for (int k = j; k > 0 && (s[k] == p[i - 1] || p[i - 1] == '.'); k--) {
                            if (dp[i - 2][k - 1]) {
                                dp[i][j] = true;
                                break;
                            }
                        }
                    }
                }
            }
        }
        return dp[N][M];
    }
};

File: Python/010. Regular Expression Matching.py
class Solution:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        m = len(s)
        n = len(p)
        s = '#' + s
        p = '#' + p
        dp = []
        for i in range(n + 1):
            dp.append([False] * (m + 1))
        dp[0][0] = True
        i = 2
        while i <= n and p[i] == '*':
            dp[i][0] = True
            i = i + 2
        for i in range(1, n + 1):
            for j in range(1, m + 1):
                if (p[i] == '.' or p[i] == s[j]) and dp[i - 1][j - 1]:
                    dp[i][j] = True
                if p[i] == '*':
                    if dp[i - 2][j]:
                        dp[i][j] = True
                    else:
                        t = j
                        while t > 0 and (p[i - 1] == s[t] or p[i - 1] == '.'):
                            if dp[i - 2][t - 1]:
                                dp[i][j] = True
                                break
                            t -= 1
                #print(i, ' ', j, ' ', dp[i][j])
        return dp[n][m]

so = Solution()
ans = so.isMatch('aab', 'c*a*b')
print(ans)